apps/prairielearn/src/batched-migrations/20230913181816_pl_courses_created_at.ts

```diff
@@ -0,0 +1,77 @@
+import execa = require('execa');
+import * as fs from 'fs-extra';
+import { makeBatchedMigration } from '@prairielearn/migrations';
+import { loadSqlEquiv, queryOneRowAsync, queryRow } from '@prairielearn/postgres';
+
+import { CourseSchema, DateFromISOString } from '../lib/db-types';
+
+const sql = loadSqlEquiv(__filename);
+
+async function getEarliestCommitDateForCourse(coursePath: string | null): Promise<Date | null> {
+  if (coursePath == null || !(await fs.pathExists(coursePath))) {
+    return null;
+  }
+
+  const res = await execa('git', ['log', '--reverse', '--format=%cI', '--date=iso'], {
+    cwd: coursePath,
+  });
+
+  const lines = res.stdout.trim().split('\n');
+  return lines.length > 0 ? new Date(lines[0]) : null;
+}
+
+async function getEarliestJobSequenceDateForCourse(courseId: string): Promise<Date | null> {
+  const result = await queryRow(
+    sql.select_earliest_job_sequence_for_course,
+    { course_id: courseId },
+    DateFromISOString.nullable(),
+  );
+
+  return result;
+}
+
+function smallestDate(first: Date | null, second: Date | null): Date | null {
+  if (first == null) {
+    return second;
+  }
+  if (second == null) {
+    return first;
+  }
+  return first < second ? first : second;
+}
+
+export default makeBatchedMigration({
+  async getParameters() {
+    const result = await queryOneRowAsync('SELECT MAX(id) as max from pl_courses;', {});
+    return {
+      min: 1n,
+      max: result.rows[0].max,
+      batchSize: 10,
+    };
+  },
+
+  async execute(start: bigint, end: bigint): Promise<void> {
+    for (let id = start; id <= end; id++) {
+      const course = await queryRow(sql.select_course, { course_id: id }, CourseSchema);
+
+      if (course.created_at != null) {
+        // The course already has a date; don't overwrite it.
+        return;
+      }
+
+      const earliestCommitDate = await getEarliestCommitDateForCourse(course.path);
+      const earliestServerJobDate = await getEarliestJobSequenceDateForCourse(course.id);
+
+      // Take the earlier of the two values.
+      const createdAt = smallestDate(earliestCommitDate, earliestServerJobDate);
+      if (!createdAt) {
+        throw new Error(`Could not determine created_at date for course ${course.id}`);
```

Comment:
I opted to fail fast if we couldn't find a date; this will give us an opportunity to manually intervene if necessary.

Are there other sources of data we could pull from to determine a rough course creation date if both of these fail?

Reply:
> Are there other sources of data we could pull from to determine a rough course creation date if both of these fail?

Earliest enrollments.created_at? Earliest page_view_logs.date? Earliest audit_logs.date?

Reply:
`enrollments.created_at` was only added a few months ago, see https://github.com/PrairieLearn/PrairieLearn/commit/55013b1602c8983828abf4197c11918d7c1baf98, so it's probably not a good candidate.

`page_view_logs.date` is an interesting possibility, though is probably pretty imprecise (a user could work for months on content without ever viewing something as a student). Maybe this is a good thing to add if we run the current migration in prod and discover that some courses are still missing dates.

The `audit_logs` table as a whole is only a last resort at the moment: it has no indexes and would thus require a full table scan to get any information out of it.

Reply:
We should always be able to determine an earliest commit date from git so I think it's safe to just fail.

Reply:
That's not necessarily true for deleted courses. If/when we add a `NOT NULL` constraint to the `created_at` column, I suppose we can always make it conditional on `deleted_at IS NOT NULL`?